Added parallax effect for cupertino fullscreenDialog

Kavantix opened the pull request:
## Description

Added the slideTransition for the parallax effect when a route was pushed as fullscreenDialog.


## Related Issues

Fixes #20073

## Tests

Added test that check that the parallax effect is working and is not used if a fullscreen routes is push over.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [ x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x ] I signed the [CLA].
- [x ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ x] I updated/added relevant documentation (doc comments with `///`).
- [x ] All existing and new tests are passing.
- [x ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x ] No, no existing tests failed, so this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

Kavantix committed:
Added parallax effect for cupertino fullscreenDialog

xster commented on packages/flutter/lib/src/cupertino/route.dart:
```diff
@@ -302,8 +302,15 @@ class CupertinoPageRoute<T> extends PageRoute<T> {
+        //
+        // In the middle of a back gesture drag, let the transition be linear to
+        // match finger motions.
+        linearTransition: isPopGestureInProgress(route),
```
I suspect this might not be needed. Can you find a native example in iOS where a fullscreen modal page is back swipable?

Kavantix replied:
Indeed the primaryanimation does not use this, however, the secondary animation does still needs it

xster replied:
ah you're right, thanks

xster commented on packages/flutter/lib/src/cupertino/route.dart:
```diff
@@ -424,27 +431,47 @@ class CupertinoFullscreenDialogTransition extends StatelessWidget {
   /// Creates an iOS-style transition used for summoning fullscreen dialogs.
   CupertinoFullscreenDialogTransition({
     Key key,
-    @required Animation<double> animation,
```
we're renaming an existing public API but I'm surprised this didn't break any tests 😅
given our latest breaking change policy, it means this one's good to go https://flutter.dev/docs/resources/compatibility

Kavantix replied:
Yeah this part of flutter isnt very well tested to be honest

xster commented on packages/flutter/test/cupertino/route_test.dart:
```diff
@@ -502,6 +502,165 @@ void main() {
     expect(tester.getTopLeft(find.byType(Placeholder)).dy, closeTo(600.0, 0.1));
   });
 
+  testWidgets('Bottom route has parallax when non fullscreen route is pushed on top', (WidgetTester tester) async {
```
fullscreen dialog route has parallax exit animation when non fullscreen route is pushed on top

Kavantix replied:
This test is not targeting full screen specifically, also its not only the exit animation that has parallax
So maybe: CupertinoPageRoute has parallax animation when non fullscreenDialog route is pushed on top

xster commented on packages/flutter/test/cupertino/route_test.dart:
```diff
@@ -502,6 +502,165 @@ void main() {
+                CupertinoButton(
+                  child: const Text('Button'),
+                  onPressed: () {
+                    Navigator.push<void>(context, CupertinoPageRoute<void>(
```
you're not testing the new functionality on CupertinoFullscreenDialogTransition you created right? That would involve 2 routes with the bottom one having fullscreenDialog true

Kavantix replied:
You’re correct, it seems I forgot to also test that explicitely.
This test is still needed though since it was missing.  
Is it considered bad or good practice to move de nody of this test into a function that gets an argument? (Whether the bottomroute should be fuulscreenDialog)

xster replied:
having helper functions is ok

Kavantix commented on packages/flutter/test/cupertino/route_test.dart:
```diff
@@ -554,6 +713,51 @@ void main() {
     expect(tester.getTopLeft(find.text('2')).dx, moreOrLessEquals(659, epsilon: 1));
   });
 
+  testWidgets('Dragged pop gesture is linear', (WidgetTester tester) async {
```
Whoops this should not be here

Kavantix committed:
Added test parallax effect of CupertinoRoute

Kavantix commented:
@xster I updated the names of the test, added that both fullscreen dialog and non fullscreen dialog cases are tested and updated the comments such that they reflect which animations are affected by the `linearTransition` boolean.  
Turns out that the comment on `CupertinoPageTransition` said that only the primary animation would be affected but that was not true so also updated that one.

Kavantix committed:
Removed trailing whitespace

xster reviewed (approved):
LGTM. Thanks for your patience!

jamesderlin commented:
Someone complained about the breakage: https://stackoverflow.com/questions/61075440/after-latest-flutter-upgrade-project-is-not-building

While I understand that this isn't strictly considered breaking according to [the breaking change policy](https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes), in this case it's clearly from a complete lack of testing.  A `@required` named parameter was changed; anyone using `CupertinoFullscreenDialogTransition` cannot avoid being broken.  This should fall under:

>  In cases where we can imagine reasonable scenarios where developers would be affected negatively, by courtesy, once the change has landed, engineers are encouraged to announce the changes by sending an e-mail to flutter-announce and listing it on our [Changelog](https://github.com/flutter/flutter/wiki/Changelog) wiki page (such that they will be included in our release notes)

xster commented:
It looks like we broke the extended_images pub package https://github.com/fluttercandies/extended_image/blob/f65ce8f5dd37b3d54d14422adf31e40a98c435f0/lib/src/gesture/extended_image_slide_page_route.dart#L331. 

@Kavantix would you like to create an announcement message and send it to flutter-announce? I'll help you approve the group message. See https://github.com/flutter/website/blob/master/src/docs/release/breaking-changes/template.md for a message template. 

I'll add this change to the changelog.

Kavantix commented:
Sure I’ll do that.

Shouldn’t the changelog also reflect that two other required parameters where added @xster?
